### PR TITLE
compiler: always allocate register to save conditional values

### LINKED ISSUE #665: failed to compile: invalid register [nil]
**Describe the bug**
Wazero fails to compile a wasm module I've compiled from Rust, which is compiled and executed successfully with Wasmtime and Wasmer.

**To Reproduce**
wasm module: https://github.com/olivierlemasle/wasm-text2png/releases/download/v0.1.0/wasm-text2png.wasm
It is a WebAssembly+WASI module compiled from Rust; its source code is in https://github.com/olivierlemasle/wasm-text2png (it simply uses the pure-Rust crate [text-to-png](https://crates.io/crates/text-to-png/) to format a text given as parameter to PNG.)
It is compiled with target `wasm32-wasi` using `cargo-wasi`.

I can compile and execute this wasm module successfully with Wasmtime and Wasmer.

My host code:

```go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/tetratelabs/wazero"
	"github.com/tetratelabs/wazero/wasi_snapshot_preview1"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Required arg: WASM file path")
		return
	}
	file := os.Args[1]

	ctx := context.Background()

	rConfig := wazero.NewRuntimeConfig().WithWasmCore2()
	r := wazero.NewRuntimeWithConfig(rConfig)
	defer r.Close(ctx)

	if _, err := wasi_snapshot_preview1.Instantiate(ctx, r); err != nil {
		log.Panicln(err)
	}

	wasm, err := os.ReadFile(file)
	if err != nil {
		log.Panicln(err)
	}

	_, err = r.CompileModule(ctx, wasm, wazero.NewCompileConfig())
	if err != nil {
		log.Panicln(err)
	}
	log.Println("Successfully compiled")
}

```

When executing with `go run . wasm-text2png.wasm`, it fails with:

```
panic: function[42/1870] failed to compile: invalid register [nil]
```

**Expected behavior**
I would expect this module to compile, as with Wasmtime or Wasmer, or fail with a more helpful message / debugging info.

**Environment (please complete the relevant information):**
- Go version: 1.18.3
- wazero Version: 2fa67b83bf6adca7b19905e5b0de7eea8e35f757
- Host architecture: amd64
- Runtime mode: compiler

### PR DESCRIPTION
**Pull request by mathetake**
Previously I had the assumption that we have always free registers for conditional value 
saving in compileLoadConditionalRegisterToGeneralPurposeRegister, but that is not the case 
when the conditional flag was created by floating-point comparison instructions.

Fixes #665 

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: internal/engine/compiler/impl_amd64.go [hunk 1]
```diff
@@ -4184,20 +4208,26 @@ func (c *amd64Compiler) compileLoadValueOnStackToRegister(loc *runtimeValueLocat
...
+	return
 }
 
 // loadConditionalRegisterToGeneralPurposeRegister saves the conditional register value
 // to a general purpose register.
-func (c *amd64Compiler) compileLoadConditionalRegisterToGeneralPurposeRegister(loc *runtimeValueLocation) {
-	// Get the free register.
-	reg, _ := c.locationStack.takeFreeRegister(registerTypeGeneralPurpose)
+func (c *amd64Compiler) compileLoadConditionalRegisterToGeneralPurposeRegister(loc *runtimeValueLocation) error {
+	reg, err := c.allocateRegister(loc.getRegisterType())
```
Comment (mathetake) [inline internal/engine/compiler/impl_amd64.go:4225]: this is the fix: previously I had the assumption that we have always free registers for conditional value saves in `compileLoadConditionalRegisterToGeneralPurposeRegister`, but that is not the case when the conditional flag was created by floating-point comparison instructions. 

This is not only the case for amd64, but also on amd64.

**Comment by mathetake** [comment]
I will backfill the unit test but this fixes the original bug already :D

**Review by codefromthecrypt — APPROVED** [review]
thanks and agreed that the extra context will help in case something similar occurs later!

**Comment by olivierlemasle** [comment]
I confirm that this fixes my issue #665, thanks.

**Comment by mathetake** [comment]
Thank you for confirming @olivierlemasle :D

**Pull request merged**